### PR TITLE
Support sf multipolygons

### PR DESCRIPTION
**Pull request by jcheng5**
This feature/refactor branch adds support for `MULTIPOLYGON` but also greatly streamlines the code needed to translate both sp and sf objects to Leaflet polygon data. See the giant comment in normalize.R, but the gist is that for any sp/sf object you only need to implement the most "natural" method among `to_ring`, `to_polygon`, `to_multipolygon`, or `to_multipolygon_list`, and rely on the default implementation of these methods to properly "box" the data the correct number of times (this was very tricky to get right previously).

`MULTIPOINT` is still not supported, but that's because I don't currently know how to implement the JS side without breaking clustering and crosstalk. At best we figure out a way to do this well, maybe by translating `MULTIPOINT` to `POINT` somewhere along the line; at worst we may need to add a new R function for adding this kind of data, that doesn't provide clustering or possibly even crosstalk.

@tim-salabim I am privately using some test cases that @timelyportfolio created using data from mapview. Do you mind if I copy the data into the leaflet package for unit testing? (I would rather copy the data than introduce a circular package dependency between mapview and leaflet)

### CODE REVIEW

**Comment by bhaskarvk** [comment]
From leaflet.extras' POV, is there any changes I need to make or new functions I need to look out for?

Thanks

**Comment by tim-salabim** [comment]
@jcheng5 sure, copy to your liking.

**Comment by jcheng5** [comment]
@bhaskarvk If you haven't implemented a `polygonData` method, then I don't think so?

**Comment by jcheng5** [comment]
Note: Fixes #355.

File: .travis.yml [hunk 1]
```diff
@@ -1,11 +1,21 @@
...
 apt_packages:
-  - r-cran-rgdal
   - libproj-dev
 
+before_install:
+  - sudo add-apt-repository ppa:ubuntugis/ubuntugis-unstable --yes
+  - sudo apt-get --yes --force-yes update -qq
+  - sudo apt-get install --yes libudunits2-dev libproj-dev libgeos-dev libgdal-dev
+
+r_packages:
```
Comment (hadley) [inline .travis.yml]: This would still be better in the DESCRIPTION

File: .travis.yml [hunk 2]
```diff
@@ -1,11 +1,21 @@
 language: r
 sudo: required
 cache: packages
+dist: trusty
 
 apt_packages:
```
Comment (hadley) [inline .travis.yml]: I think you can drop this given that you're installing `libproj-dev` below

File: R/normalize-sf.R [hunk 3]
```diff
@@ -23,44 +23,46 @@ pointData.sfc_POINT <- function(obj) {
   )
 }
 
+#' @export
+pointData.POINT <- function(obj) {
+  check_crs(obj)
+
+  bbox <- sf_bbox(obj)
+
+  if (!is.matrix(obj)) {
```
Comment (hadley) [inline R/normalize-sf.R]: When would this not be a matrix? A comment seems useful

File: R/normalize-sf.R [hunk 4]
```diff
@@ -23,44 +23,46 @@ pointData.sfc_POINT <- function(obj) {
   )
 }
 
+#' @export
+pointData.POINT <- function(obj) {
+  check_crs(obj)
+
+  bbox <- sf_bbox(obj)
```
Comment (hadley) [inline R/normalize-sf.R]: Just inline in `sf_bbox()`?

File: R/normalize.R [hunk 5]
```diff
@@ -104,7 +104,7 @@ pointData <- function(obj) {
 #' @export
 pointData.default <- function(obj) {
   stop("Don't know how to get location data from object of class ",
-    class(obj)[[1]])
+    paste(class(obj), collapse = ","))
```
Comment (hadley) [inline R/normalize.R]: `", "`?

File: R/normalize.R [hunk 6]
```diff
@@ -165,3 +165,80 @@ checkMatrix <- function(x) {
...
+# - to_ring
+#
+# Each of the specific sp/sf classes need only implement whichever ONE of those
+# actually makes sense (e.g. to_multipolygon_list.sfc,
+# to_multipolygon.MULTIPOLYGON, to_polygon.POLYGON, to_ring.LINESTRING). The
+# higher-level polygonData wrappers will simply call to_multipolygon_list(x),
+# and the default implementations of those methods will fall through to the next
+# level until a match is found.
+
+to_multipolygon_list <- function(x) {
```
Comment (hadley) [inline R/normalize.R]: If you want this to be extensible from outside leaflet, you'll need to export the generics too.

Comment (tim-salabim) [inline R/normalize.R]: exporting would highly be appreciated by selected developers that build their efforts on leaflet :-)

**Comment by tim-salabim** [comment]
I am guessing their not in data because they are actually borrowed from mapview-package which has leaflet as a dependency. However, we could migrate them to leaflet and delete them from mapview if that is helping.

**Comment by jcheng5** [comment]
@tim-salabim That's correct. I'm open to either, just let me know your preference.

File: .travis.yml (hunk 1 above)
Comment (jcheng5) [inline .travis.yml]: Sorry, can you elaborate? I'm pretty inexperienced when it comes to Travis. These packages are all in the DESCRIPTION under Suggests, but don't I need them listed here to get them to install for testing?

Comment (hadley) [inline .travis.yml]: Any packages in DESCRIPTION are automatically installed

File: R/normalize-sf.R (hunk 4 above)
Comment (jcheng5) [inline R/normalize-sf.R]: I have to do this because I may turn it into a matrix below and that drops the bbox

**Comment by tim-salabim** [comment]
My cunning plan for mapview is to focus my efforts on sf classes rather than sp. Would it make sense to have these sp data in leaflet and mapview provide sf versions of those?

File: .travis.yml (hunk 1 above)
Comment (tim-salabim) [inline .travis.yml]: I think sf (or some of its depends/imports, e.g. udunits2) specifically needs the dev version of these system libraries to be available and I don't think travis installs those by default.

Comment (hadley) [inline .travis.yml]: @tim-salabim we're talking about R packages here, not system libraries

Comment (tim-salabim) [inline .travis.yml]: @hadley I thought you're talking about the sudo apt-get calls, especially the one in line 12. If you're not, just ignore my comment

Comment (bhaskarvk) [inline .travis.yml]: Agree with @hadley, you don't need the r_packages section here.

File: R/normalize.R (hunk 6 above)
Comment (jcheng5) [inline R/normalize.R]: @tim-salabim You want to call it, or extend it? I don't think you want to call it--it returns very specific nested list structures that are only relevant to leaflet, most likely. And you don't want to implement it either--`derivePolygons` is what you want to implement. These `to_xxx` generics are just to make the implementation of `derivePolygons` less confusing for me, I don't think they make sense as part of the public contract of the package.

If you have strong feelings or a specific need that doesn't reflect the above, please let me know, I'm happy to reconsider.

Comment (tim-salabim) [inline R/normalize.R]: ping @timelyportfolio as this may be relevant for mapedit

Comment (timelyportfolio) [inline R/normalize.R]: @tim-salabim,@jcheng5 unexported should be fine.  I don't think we will need these in `mapedit`.  Thanks for all your work on this.  `sf` is definitely the future!

Comment (tim-salabim) [inline R/normalize.R]: @jcheng5 then no export needed. Cheers

**Pull request merged**